docs/using/syntax-optional.md

```diff
@@ -1,32 +1,221 @@
+
+`www.example.com` -> www.example.com
+
+:::{important}
+This extension requires that [linkify-it-py](https://github.com/tsutsu3/linkify-it-py) is installed.
+Either directly; `pip install linkify-it-py` or *via* `pip install myst-parser[linkify]`.
+:::
+
+(syntax/substitutions)=
+
+## Substitutions (with Jinja2)
+
+Adding `"substitution"` to `myst_enable_extensions` (in the sphinx `conf.py` [configuration file](https://www.sphinx-doc.org/en/master/usage/configuration.html)) will allow you to add substitutions, added in either the `conf.py` using `myst_substitutions`:
+
+```python
+myst_substitutions = {
+  "key1": "I'm a **substitution**"
+}
+```
+
+or at the top of the file, in the front-matter section (see [this section](syntax/frontmatter)):
+
+````yaml
+---
+substitutions:
+  key1: "I'm a **substitution**"
+  key2: |
+    ```{note}
+    {{ key1 }}
+    ```
+  key3: <img src="img/fun-fish.png" alt="fishy" width="200px">
+---
+````
 
-A special syntax for admonitions can optionally be enabled by setting `myst_admonition_enable = True` in the sphinx `conf.py` [configuration file](https://www.sphinx-doc.org/en/master/usage/configuration.html).
+:::{important}
+Keys in the front-matter will override ones in the `conf.py`.
+:::
 
-The key differences are that, instead of back-ticks, colons are used, and **the content starts as regular Markdown**.
-This has the benefit of allowing the content to be rendered correctly, when you are working in any standard Markdown editor.
-For example:
+You can use these substitutions inline or as blocks, and you can even nest substitutions in other substitutions (but circular references are prohibited):
+
+:::{tabbed} Markdown Input
 
+```md
+Inline: {{ key1 }}
+
+{{ key2 }}
+
+| col1     | col2     |
+| -------- | -------- |
+| {{key3}} | {{key3}} |
+
+```
+
+:::
+
+:::{tabbed} Rendered Output
+Inline: {{ key1 }}
+
+{{ key2 }}
+
+| col1     | col2     |
+| -------- | -------- |
+| {{key3}} | {{key3}} |
+
+:::
+
+:::{important}
+
+Substitutions will only be assessed where you would normally use Markdown, e.g. not in code blocks:
+
+````
+```
+{{ key1 }}
+```
+````
+
+```
+{{ key1 }}
+```
+
+Also, it should be noted that, inline substitutions will not parse block-level syntaxes like directives:
+
+```
+inline {{ key2 }}
+```
+
+inline {{ key2 }}
+
+:::
+
+Substitution references are assessed as [Jinja2 expressions](http://jinja.palletsprojects.com) which can use [filters](https://jinja.palletsprojects.com/en/2.11.x/templates/#list-of-builtin-filters), and also contains the [Sphinx Environment](https://www.sphinx-doc.org/en/master/extdev/envapi.html) in the context (as `env`).
+Therefore you can do things like:
+
+```md
+{{ env.docname | upper }}
+{{ "a" + "b" }}
+```
+
+{{ env.docname | upper }}
+{{ "a" + "b" }}
```

Comment:
This makes me wonder if there is a way to (in the future, not this PR) incorporate glue-like functionality into this syntax. I think `{{ mykey }}` would be the natural short-hand for ` {glue:}`mykey` `, don't you think? Would using this syntax *both* for "markdown substitutions" **and** for glue-like syntax be possible? (again, I do not think we should do it in this PR, just checking whether it would be possible)

Reply:
Oh yeh this was exactly in my thinking

Reply:
nice - I think using Jinja for this is super cool :-)